cadquery/cq_directive.py

```diff
@@ -84,9 +114,95 @@ def run(self):
         return []
 
 
+class cq_directive_vtk(Directive):
+
+    has_content = True
+    required_arguments = 0
+    optional_arguments = 2
+    option_spec = {
+        "height": directives.length_or_unitless,
+        "width": directives.length_or_percentage_or_unitless,
+        "align": directives.unchanged,
+    }
+
+    def run(self):
+
+        options = self.options
+        content = self.content
+        state_machine = self.state_machine
+        env = self.state.document.settings.env
+        build_path = Path(env.app.builder.outdir)
+        out_path = build_path / "_static"
+
+        # only consider inline snippets
+        plot_code = "\n".join(content)
+
+        # collect the result
+        try:
+            result = cqgi.parse(plot_code).build()
+
+            if result.success:
+                if result.first_result:
+                    shape = result.first_result.shape
+                else:
+                    shape = result.env["result"]
+
+                if isinstance(shape, Assembly):
+                    assy = shape
+                else:
+                    assy = Assembly(shape, color=Color(*DEFAULT_COLOR))
+            else:
+                raise result.exception
+
+        except Exception:
+            traceback.print_exc()
+            assy = Assembly(Compound.makeText("CQGI error", 10, 5))
+
+        # save vtkjs to static
+        fname = Path(str(uuid()))
+        exporters.assembly.exportVTKJS(assy, out_path / fname)
+        fname = str(fname) + ".zip"
+
+        # add the output
+        lines = []
+
+        data = dumps(toJSON(assy))
+
+        lines.extend(
+            template_vtk.format(
+                code=indent(TEMPLATE_RENDER.format(), "    "),
+                data=data,
+                ratio="null",
+                element="document.currentScript.parentNode",
+                txt_align=options.get("align", "left"),
+                width=options.get("width", "100%"),
+                height=options.get("height", "500px"),
+            ).splitlines()
+        )
+
+        lines.extend(["::", ""])
+        lines.extend(["    %s" % row.rstrip() for row in plot_code.split("\n")])
+        lines.append("")
+
+        if len(lines):
+            state_machine.insert_input(lines, state_machine.input_lines.source(0))
+
+        return []
+
+
 def setup(app):
     setup.app = app
     setup.config = app.config
     setup.confdir = app.confdir
 
     app.add_directive("cq_plot", cq_directive)
+    app.add_directive("cq_vtk", cq_directive_vtk)
+
+    # download and add vtk.js
+    build_path = Path(app.outdir)
+    out_path = build_path / "_static"
+    out_path.mkdir(parents=True, exist_ok=True)
+
+    request.urlretrieve("https://unpkg.com/vtk.js", out_path / "vtk.js")
```

Comment:
Is there any alternative to this? The Nix package manager does not allow network access during build time (because it produces deterministic builds). I'm pretty sure this is the first bit of network access (outside of package managers) in the CQ code base.

I can download the file before building and patch this on the Nix side if there is no reasonable alternative.

Reply:
I can bundle the file. BTW: Shouldn't `cq_directive` be moved completely to`doc`?

Reply:
> I can bundle the file.

Thanks!

> Shouldn't cq_directive be moved completely todoc?

I agree, the Sphinx docs seem to suggest that structure.